lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -125,83 +138,177 @@ impl Client {
+                    DeviceControlResponse::new(
+                        &ioctl,
+                        NTSTATUS::STATUS_SUCCESS.to_u32().unwrap(),
+                        vec![],
+                    )
+                };
+                debug!("replying with: {:?}", resp);
+                let resp = self.add_headers_and_chunkify(
+                    PacketId::PAKID_CORE_DEVICE_IOCOMPLETION,
+                    resp.encode()?,
+                )?;
+                debug!("sending device IO response");
+                Ok(resp)
             }
-            let resp = self.add_headers_and_chunkify(
-                consts::PacketId::PAKID_CORE_DEVICE_IOCOMPLETION,
-                DeviceControlResponse::new(&ioctl, code, res).encode()?,
-            )?;
-            debug!("sending device IO response");
-            Ok(resp)
-        } else {
-            Err(invalid_data_error(&format!(
+            _ => Err(invalid_data_error(&format!(
+                // TODO(isaiah): send back a not implemented response(?)
                 "got unsupported major_function in DeviceIoRequest: {:?}",
-                &req.major_function
-            )))
+                &major_function
+            ))),
         }
     }
 
+    pub fn write_client_device_list_announce<S: Read + Write>(
+        &mut self,
+        req: ClientDeviceListAnnounce,
+        mcs: &mut mcs::Client<S>,
+    ) -> RdpResult<()> {
+        self.push_active_device_id(req.device_list[0].device_id)?;
+        debug!("sending new drive for redirection: {:?}", req);
+
+        let responses =
+            self.add_headers_and_chunkify(PacketId::PAKID_CORE_DEVICELIST_ANNOUNCE, req.encode()?)?;
+        let chan = &CHANNEL_NAME.to_string();
+        for resp in responses {
+            mcs.write(chan, resp)?;
+        }
+
+        Ok(())
+    }
+
     /// add_headers_and_chunkify takes an encoded PDU ready to be sent over a virtual channel (payload),
     /// adds on the Shared Header based the passed packet_id, adds the appropriate (virtual) Channel PDU Header,
     /// and splits the entire payload into chunks if the payload exceeds the maximum size.
     fn add_headers_and_chunkify(
         &self,
-        packet_id: consts::PacketId,
+        packet_id: PacketId,
         payload: Vec<u8>,
     ) -> RdpResult<Vec<Vec<u8>>> {
-        let mut inner =
-            SharedHeader::new(consts::Component::RDPDR_CTYP_CORE, packet_id).encode()?;
+        let mut inner = SharedHeader::new(Component::RDPDR_CTYP_CORE, packet_id).encode()?;
         inner.extend_from_slice(&payload);
         self.vchan.add_header_and_chunkify(None, inner)
     }
+
+    fn push_active_device_id(&mut self, device_id: u32) -> RdpResult<()> {
+        if self.active_device_ids.contains(&device_id) {
+            return Err(RdpError::TryError(format!(
+                "attempted to add a duplicate device_id {} to active_device_ids {:?}",
```

Comment:
This will effectively only come due to a coding error on my part. For example, if the client tries to share a directory with device id set to `SCARD_DEVICE_ID`, or if it tries to share the same directory twice with the same ID, or tries to share two different directories with the same id (etc.).

I think its ok to throw it at the user, since it should never happen, and if it does it will be helpful to know if they contact us for debugging. I'm not sure how we typically handle such a case though.